### PR TITLE
stop filtering "phantom archives" when it won't work; reenable it on request

### LINKED ISSUE #4511: "phantom archive" filtering for reconnected fetch-by-key streams
#4509 implies that a reconnected fetch-by-key stream won't have the state to filter out phantom archives, as a stream from the ledger-begin does. In #4509 we continue filtering out phantom archives iff an offset *is not* specified.

@hurryabit , @gerolf-da to determine what kind of burden is acceptable to place on users to correctly pass in a form of the fetch stream's state when reconnecting and what that data ought to look like. We also need to decide how to explain the behavior if the data passed in is wrong, besides the usual [GIGO](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out).

Depends on #4509.

### PR DESCRIPTION
**Pull request by S11001001**
Filtering was enabled for resumed key streams even though this was wrong; this restores the proper behavior if you don't specify hints, and allows you to specify hints if you want to reenable filtering. See the documentation for details. Fixes #4511.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: docs/source/json-api/index.rst [hunk 1]
```diff
@@ -1302,3 +1302,27 @@ Example:
...
+contract but no `"def345"` contract, you might specify:
+
+.. code-block:: json
+
+    [
+        {"templateId": "Account:Account", "key": {"_1": "Alice", "_2": "abc123"},
+         "offsetHint": "#1:0"},
+        {"templateId": "Account:Account", "key": {"_1": "Alice", "_2": "def345"},
+         "offsetHint": null}
+    ]
```
Comment (leo-da) [inline docs/source/json-api/index.rst]: @hurryabit we should rename `offsetHint` to `contractIdHint` or `contractIdOffsetHint`.

Remember we have offset ticks defined as:
```
{"events":[],"offset":"5609"}
```
The `offset` value from the tick and `offsetHint` are two different things.

Comment (leo-da) [reply docs/source/json-api/index.rst]: let's merge it... and discuss the field name during the catch-up meeting.

Comment (hurryabit) [reply docs/source/json-api/index.rst]: Yes, I think `contractIdHint` (or maybe `contractIdAtOffset`?) would be a better name.

File: docs/source/json-api/index.rst [hunk 2]
```diff
@@ -1302,3 +1302,27 @@ Example:
...
 
 The output stream has the same format as the output from the `Contracts Query Stream`_. We further guarantee that for every ``archived`` event appearing on the stream there has been a matching ``created`` event earlier in the stream.
+
+You may supply an optional ``offset`` for the stream, exactly as with
+query streams.  However, you must supply with each ``{templateId,
+key}`` pair an ``offsetHint``, which is the contract ID currently
+associated with that pair at the point of the given offset, or
+``null`` if no contract ID was associated with the pair at that
+offset.  For example, with the above keys, if you had one ``"abc123"``
+contract but no `"def345"` contract, you might specify:
```
Comment (hurryabit) [inline docs/source/json-api/index.rst]:
```suggestion
contract but no ``"def345"`` contract, you might specify:
```

File: docs/source/json-api/index.rst [hunk 3]
```diff
@@ -1302,3 +1302,27 @@ Example:
...
+        {"templateId": "Account:Account", "key": {"_1": "Alice", "_2": "def345"},
+         "offsetHint": null}
+    ]
+
+If every ``offsetHint`` is specified, as is so in the example above,
+you will not receive any ``archived`` events for contracts created
+before the offset *unless* those contracts are identified in an
+``offsetHint``.  By contrast, if any ``offsetHint`` is missing,
+``archived`` event filtering will be disabled, and you will receive
+"phantom archives" as with query streams.
```
Comment (hurryabit) [inline docs/source/json-api/index.rst]: Is there an explanation of what "phantom archives" are elsewhere?

Comment (S11001001) [reply docs/source/json-api/index.rst]: [Yes, above.](https://github.com/digital-asset/daml/commit/fb4284595933a22c74304df39d64d6ccb78b5589#diff-acc1415ff18516eee3cce71dc3407d5cR1212-R1218)

File: docs/source/json-api/index.rst (hunk 2 above)
Comment (S11001001) [reply docs/source/json-api/index.rst]: 811a1a9952, thanks

File: docs/source/json-api/index.rst (hunk 1 above)
Comment (S11001001) [reply docs/source/json-api/index.rst]: > `contractIdAtOffset`

What we've decided on.

**Pull request merged**